client/src/components/graph/overlays/centroidLabels.js

```diff
@@ -2,42 +2,114 @@ import React, { PureComponent } from "react";
 import { connect } from "react-redux";
 
 import { categoryLabelDisplayStringLongLength } from "../../../globals";
+import calcCentroid from "../../../util/centroid";
+import { createColorQuery } from "../../../util/stateManager/colorHelpers";
 
 export default
 @connect((state) => ({
-  colorAccessor: state.colors.colorAccessor,
+  annoMatrix: state.annoMatrix,
+  colors: state.colors,
+  layoutChoice: state.layoutChoice,
   dilatedValue: state.pointDilation.categoryField,
-  labels: state.centroidLabels.labels,
   categoricalSelection: state.categoricalSelection,
+  showLabels: state.centroidLabels?.showLabels,
 }))
 class CentroidLabels extends PureComponent {
-  // Check to see if centroids have either just been displayed or removed from the overlay
+  constructor(props) {
+    super(props);
+    this.state = {
+      status: "pending",
+      labels: new Map(),
+      colorAccessor: null,
+    };
+  }
+
+  componentDidMount() {
+    this.updateState(null);
+  }
 
   componentDidUpdate(prevProps) {
-    const { labels, overlayToggled } = this.props;
-    const prevSize = prevProps.labels.size;
-    const { size } = labels;
+    this.updateState(prevProps);
 
-    const displayChangeOff = prevSize > 0 && size === undefined;
-    const displayChangeOn = prevSize === undefined && size > 0;
+    const { showLabels, overlaySetShowing } = this.props;
+    const { labels } = this.state;
+    overlaySetShowing("centroidLabels", showLabels && labels.size > 0);
+  }
+
+  colorByQuery() {
+    const { annoMatrix, colors } = this.props;
+    const { schema } = annoMatrix;
+    const { colorMode, colorAccessor } = colors;
+    return createColorQuery(colorMode, colorAccessor, schema);
+  }
 
-    if (displayChangeOn || displayChangeOff) {
-      // Notify overlay layer of display change
-      overlayToggled("centroidLabels", displayChangeOn);
+  async fetchData() {
+    const { annoMatrix, layoutChoice } = this.props;
+    // fetch all data we need: layout, category
+    const promises = [];
+    // layout
+    promises.push(annoMatrix.fetch("emb", layoutChoice.current));
+    // category to label - we ONLY label on obs, never on X, etc.
+    const query = this.colorByQuery();
+    if (query && query[0] === "obs") {
+      promises.push(annoMatrix.fetch(...query));
+    } else {
+      promises.push(Promise.resolve(null));
```

Comment:
I did it for clarity, and because I thought `null` was a better (explicit) return.  But if the caller treats null and undefined as equivalent, then you are right.  Just a style thing....